### PR TITLE
Fix #6640 Visualization issue with Waterfall plot when axes are swapped

### PR DESCRIPTION
**Pull request by ngocnn1104**
# Fix [#6640](https://github.com/cBioPortal/cbioportal/issues/6640) Visualization issue with Waterfall plot when axes are swapped

## Code changes:
- Removed offsetY and invertAxis props from VictoryAxis to make use of domainPadding prop in VictoryChart
- Sorted plot data in horizontal mode in descending order to invert the axis
- Adjusted the tooltip's position to display it more correctly
- Though it is not mentioned in this issue, I also fixed the overflow of tooltip content

### Before:
Screenshot from 2019-10-03 14-57-39

Screenshot from 2019-10-03 15-48-04

### After:
Screenshot from 2019-10-03 14-58-24

Screenshot from 2019-10-03 15-46-36

### CODE REVIEW

File: src/shared/components/plots/WaterfallPlot.tsx [hunk 1]
```diff
@@ -72,6 +72,10 @@ const limitValueAppearanceSupplement = {
     strokeOpacity: 1,
     size: 3
 };
+const plotTitleStyles = {
+    textAlign: 'center' as 'center',
+    margin: '20px 0 10px 0'
+};
```
Comment (pvannierop) [inline src/shared/components/plots/WaterfallPlot.tsx]: Better to put this in a `style.scss` file in the src/shared/components/plots, import it and apply via classname.

File: src/shared/components/plots/WaterfallPlot.tsx [hunk 2]
```diff
@@ -104,7 +108,9 @@ export default class WaterfallPlot<D extends IBaseWaterfallPlotData> extends Rea
                                 if (this.props.horizontal && !props.datum.searchLabel) {
                                     const x = props.x;
                                     props.x = props.y;
-                                    props.y = x;
+                                    props.y = x + 38; // to position the tooltip more correctly
+                                } else if (!this.props.horizontal) {
+                                    props.y += 25; // to position the tooltip more correctly
```
Comment (pvannierop) [inline src/shared/components/plots/WaterfallPlot.tsx]: Can you capture these hard coded values as constants at the top of the script?

File: src/shared/components/plots/WaterfallPlot.tsx [hunk 3]
```diff
@@ -454,11 +461,18 @@ export default class WaterfallPlot<D extends IBaseWaterfallPlotData> extends Rea
     
     @bind
     private getChart() {
+        const svgViewBox = () => {
+            if (this.props.horizontal) {
+                return `0 -10 ${this.svgWidth} ${this.svgHeight}`;
+            }
+            return `0 0 ${this.svgWidth} ${this.svgHeight}`
+        };
```
Comment (pvannierop) [inline src/shared/components/plots/WaterfallPlot.tsx]: Better move this out of the getCart() method and change into a @computed field that depends on this.props.horizontal?

File: src/shared/components/plots/WaterfallPlot.tsx [hunk 4]
```diff
@@ -482,12 +496,11 @@ export default class WaterfallPlot<D extends IBaseWaterfallPlotData> extends Rea
                             domainPadding={PLOT_DATA_PADDING_PIXELS}
                             singleQuadrantDomainPadding={false}
                         >
-                            {this.title}
                             {this.legend}
                             {this.props.horizontal && <VictoryAxis
                                 domain={this.plotDomainX}
                                 orientation="top"
-                                offsetY={80}
+                                offsetY={30}
```
Comment (pvannierop) [inline src/shared/components/plots/WaterfallPlot.tsx]: I know that it was hard coded before, but better extract this as a constant at the top of the screen under a clear name.

File: src/shared/components/plots/WaterfallPlot.tsx [hunk 5]
```diff
@@ -229,7 +235,8 @@ export default class WaterfallPlot<D extends IBaseWaterfallPlotData> extends Rea
 
     @computed get plotDomainY() {
         if (this.props.horizontal) {
-            return this.plotDomain.order;
+            // extend the max domain of y-axis to make up for the offset of x-axis
+            return [...this.plotDomain.order, this.plotDomain.order[this.plotDomain.order.length - 1] + 10];
```
Comment (pvannierop) [inline src/shared/components/plots/WaterfallPlot.tsx]: Better move this logic to the  `plotDomain()` method.

File: src/shared/components/plots/WaterfallPlot.tsx [hunk 6]
```diff
@@ -454,11 +461,18 @@ export default class WaterfallPlot<D extends IBaseWaterfallPlotData> extends Rea
...
+                return `0 -10 ${this.svgWidth} ${this.svgHeight}`;
+            }
+            return `0 0 ${this.svgWidth} ${this.svgHeight}`
+        };
         return (
             <div
                 ref={this.containerRef}
                 style={{width: this.svgWidth, height: this.svgHeight}}
             >
+                <p style={plotTitleStyles}>{this.title}</p>
```
Comment (pvannierop) [inline src/shared/components/plots/WaterfallPlot.tsx]: Why did you move the title out of the svg component? This may (I do not know) give a problem with export of the figure?

File: src/shared/components/plots/WaterfallPlot.tsx [hunk 7]
```diff
@@ -220,9 +223,13 @@ export default class WaterfallPlot<D extends IBaseWaterfallPlotData> extends Rea
         let max = _(this.waterfallPlotData).map('value').max() || 0;
         let min = _(this.waterfallPlotData).map('value').min() || 0;
 
+        // return range defined by the number of samples for the x-axis,
+        // extend the max domain of y-axis to make up for the offset of x-axis in horizontal mode
+        const order = this.props.horizontal ? [1, this.waterfallPlotData.length] : [1, this.waterfallPlotData.length, this.waterfallPlotData.length+10];
+
```
Comment (pvannierop) [inline src/shared/components/plots/WaterfallPlot.tsx]: 1] Can you also extract the `10` into a constant with a name that provides a clue to what this value does for the plot?

2] I suspect that modification of the plot domain is there to allow for space to shift the plot down in horizontal view. From the docs of victory chart it tells about the`plotDomain`:

>The domain prop describes the range of data the component will include. This prop can be given as a array of the minimum and maximum expected values of the data or as an object that specifies separate arrays for x and y. If this prop is not provided, a domain will be calculated from data, or other available information.

Since we do not need to accomodate new values, I think this approach is not te best way to shift the graph down. Let's discuss when you are back at the office?

Comment (pvannierop) [reply src/shared/components/plots/WaterfallPlot.tsx]: I think you better use `domainPadding` (https://formidable.com/open-source/victory/docs/common-props/#domainpadding)

Comment (ngocnn1104) [reply src/shared/components/plots/WaterfallPlot.tsx]: > I think you better use `domainPadding` (https://formidable.com/open-source/victory/docs/common-props/#domainpadding)

The domain padding is already there but somehow it doesn't work as expected. I will try to figure out why and we will discuss tomorrow.

File: src/pages/resultsView/plots/PlotsTabUtils.tsx [hunk 8]
```diff
@@ -1362,7 +1362,13 @@ function generalWaterfallPlotTooltip<D extends IWaterfallPlotData>(
 
     return (
         <div>
-            <a href={getSampleViewUrl(d.studyId, d.sampleId)} target="_blank">{d.sampleId}</a>
+            <a
+                href={getSampleViewUrl(d.studyId, d.sampleId)}
+                target="_blank"
+                style={{whiteSpace: 'pre-line', wordBreak: 'break-word'}}
```
Comment (alisman) [inline src/pages/resultsView/plots/PlotsTabUtils.tsx]: why do we want to break in the middle of a sampleId?

Comment (alisman) [reply src/pages/resultsView/plots/PlotsTabUtils.tsx]: nevermind.  i see in screenshot.  thanks!

File: src/shared/components/plots/WaterfallPlot.tsx [hunk 9]
```diff
@@ -229,7 +243,7 @@ export default class WaterfallPlot<D extends IBaseWaterfallPlotData> extends Rea
 
     @computed get plotDomainY() {
         if (this.props.horizontal) {
-            return this.plotDomain.order;
+            this.plotDomain.order;
```
Comment (adamabeshouse) [inline src/shared/components/plots/WaterfallPlot.tsx]: missing return

File: src/shared/components/plots/WaterfallPlot.tsx [hunk 10]
```diff
@@ -471,23 +488,22 @@ export default class WaterfallPlot<D extends IBaseWaterfallPlotData> extends Rea
...
                         >
                         <VictoryChart
                             theme={CBIOPORTAL_VICTORY_THEME}
                             width={this.props.chartWidth}
                             height={this.props.chartHeight}
                             standalone={false}
-                            domainPadding={PLOT_DATA_PADDING_PIXELS}
-                            singleQuadrantDomainPadding={false}
+                            domainPadding={PLOT_DATA_PADDING_PIXELS(this.props.horizontal)}
+                            singleQuadrantDomainPadding={this.props.horizontal ? true : false}
```
Comment (adamabeshouse) [inline src/shared/components/plots/WaterfallPlot.tsx]: `singleQuadrantDomainPadding={this.props.horizontal}`

File: src/shared/components/plots/WaterfallPlot.tsx [hunk 11]
```diff
@@ -60,18 +60,27 @@ const DEFAULT_FONT_FAMILY = "Verdana,Arial,sans-serif";
 export const LEGEND_Y = 30
 const RIGHT_PADDING = 120; // room for correlation info and legend
 const NUM_AXIS_TICKS = 8;
-const PLOT_DATA_PADDING_PIXELS = {y: 3, x: [30, 30]};
+const PLOT_DATA_PADDING_PIXELS = (horizontal:Boolean) => {
+    if (horizontal) {
+        return 30;
+    }
+    return {y: 3, x: 30};
+}
```
Comment (pvannierop) [inline src/shared/components/plots/WaterfallPlot.tsx]: This is not a constant. No captalization should be used here. This should be refactored into a @computed field `dataPaddingInPixels` as:
```
@computed get dataPaddingInPixels() {
   if (this.props.horizontal) {
        return 30;
    }
    return {y: 3, x: 30};
}
```

Also, this method returns different data types depending on the value of `horizontal`. Is this supported by TypeScript? I guess you have to add a return type and refactor to have the code handle it.

**Comment by alisman** [comment]
Hi guys, sorry I didn't mention this before.   Instead of merging master into branch as was done here, we should rebase these changes on master.  And while we're at it, we can squash the 6 legit commits to a single one.

**Comment by ngocnn1104** [comment]
> @ngocnn1104 Hi guys, sorry I didn't mention this before. Instead of merging master into branch as was done here, we should rebase these changes on master. And while we're at it, we can squash the 6 legit commits to a single one.

This branch has been rebased onto master and the multiple commits were squashed.
